File: Assets/Script/Channel/ChannelManager.cs
﻿using System;
using GameTools.Singletons;
using XLua;

namespace Common.GameChannel
{
    
    /// <summary>
    /// 通道管理程序
    /// </summary>
    [Hotfix]
    [LuaCallCSharp]
    public class ChannelManager:MMOSingleton<ChannelManager>
    {
        private BaseChannel channel = null;
        private Action onInitCompleted = null;
        private Action onActionSucceed = null;
        private Action onActionFailed = null;
        private Action<int> onActionProgressValueChange = null;
        public string channelName
        {
            get;
            protected set;
        }
        
        public string noticeVersion
        {
            get;
            set;
        }

        public string resVersion
        {
            get;
            set;
        }

        public string appVersion
        {
            get;
            set;
        }

        public string svnVersion
        {
            get;
            set;
        }
        public void Init(string channelName)
        {
            this.channelName = channelName;
            channel = CreateChannel(channelName);

            // AndroidSDKListener.Instance.Startup();
        }
        public bool IsInternalVersion()
        {
            return channel == null || channel.IsInternalChannel();
        }

        public BaseChannel CreateChannel(string channelName)
        {
            ChannelType platName = (ChannelType)Enum.Parse(typeof(ChannelType), channelName);
            // switch ((platName))
            // {
            //     case ChannelType.Test:
            //         return new TestChannel();
            //     default:
            //         return new TestChannel();
            // }
            return null;
        }
        public void StartDownloadGame(string url, Action succeed = null, Action fail = null, Action<int> progress = null, string saveName = null)
        {
            onActionSucceed = succeed;
            onActionFailed = fail;
            onActionProgressValueChange = progress;
            channel.DownloadGame(url, saveName);
        }
    }
}

File: Assets/.lpproj/Apis/Unity/CS/Common/GameChannel/ChannelManager.lua
---@class CS.Common.GameChannel.ChannelManager : CS.GameTools.Singletons.MMOSingleton
CS.Common.GameChannel.ChannelManager = {}

---@property readwrite CS.Common.GameChannel.ChannelManager.channelName : CS.System.String
CS.Common.GameChannel.ChannelManager.channelName = nil

---@property readwrite CS.Common.GameChannel.ChannelManager.noticeVersion : CS.System.String
CS.Common.GameChannel.ChannelManager.noticeVersion = nil

---@property readwrite CS.Common.GameChannel.ChannelManager.resVersion : CS.System.String
CS.Common.GameChannel.ChannelManager.resVersion = nil

---@property readwrite CS.Common.GameChannel.ChannelManager.appVersion : CS.System.String
CS.Common.GameChannel.ChannelManager.appVersion = nil

---@property readwrite CS.Common.GameChannel.ChannelManager.svnVersion : CS.System.String
CS.Common.GameChannel.ChannelManager.svnVersion = nil

---@return CS.Common.GameChannel.ChannelManager
function CS.Common.GameChannel.ChannelManager()
end

---@param channelName : CS.System.String
function CS.Common.GameChannel.ChannelManager:Init(channelName)
end

---@return CS.System.Boolean
function CS.Common.GameChannel.ChannelManager:IsInternalVersion()
end

---@param channelName : CS.System.String
---@return CS.Common.GameChannel.BaseChannel
function CS.Common.GameChannel.ChannelManager:CreateChannel(channelName)
end

---@param url : CS.System.String
---@param succeed : CS.System.Action
---@param fail : CS.System.Action
---@param progress : CS.System.Action
---@param saveName : CS.System.String
function CS.Common.GameChannel.ChannelManager:StartDownloadGame(url, succeed, fail, progress, saveName)
end

---@return CS.System.String
function CS.Common.GameChannel.ChannelManager:GetProductName()
end

---@param succeed : CS.System.Action
---@param fail : CS.System.Action
function CS.Common.GameChannel.ChannelManager:InstallGame(succeed, fail)
end